step-automation-packages/step-automation-packages-controller/src/main/java/step/automation/packages/AutomationPackageExecutionPlugin.java

```diff
@@ -30,40 +32,68 @@ public AutomationPackageExecutionPlugin(AutomationPackageLocks automationPackage
+                    debugLog(context, "Timeout while acquiring read lock on automation package.");
                     throw new PluginCriticalException("Timeout while acquiring lock on automation package with id " +
                             automationPackageId + ". This usually means that an update of this automation package is on-going and took more than the property " +
                             AUTOMATION_PACKAGE_READ_LOCK_TIMEOUT_SECS + " (default " + AUTOMATION_PACKAGE_READ_LOCK_TIMEOUT_SECS_DEFAULT + " seconds)");
                 }
-                logger.debug("Acquired read lock on automation package " + automationPackageId);
+                debugLog(context,"Acquired read lock on automation package.");
             } catch (InterruptedException e) {
                 throw new PluginCriticalException("Thread interrupted while acquiring lock on automation package with id " + automationPackageId);
             }
         }
     }
 
+    private void debugLog(ExecutionContext context, String message) {
+        if (logger.isDebugEnabled()) {
+            Plan plan = context.getPlan();
+            String planName = (plan != null) ? plan.getAttribute(AbstractOrganizableObject.NAME) : "unresolved";
+            String automationPackageId = getAutomationPackageId(context);
+            logger.debug(message + " Id: " + automationPackageId + ", execution id: " +
+                    context.getExecutionId() + ",  plan: " + planName);
+        }
+    }
+
+    public boolean shouldLock(ExecutionContext context) {
+        String automationPackageId = getAutomationPackageId(context);
+        if (automationPackageId == null) {
+            return false;
+        } else {
+            Object byPassId = context.getExecutionParameters().getCustomField(BYPASS_AUTOMATION_PACKAGE_LOCK_FOR_ID);
+            if (byPassId == null) {
+                return true;
+            } else if (byPassId instanceof String) {
+                //Bypass lock if same automation package is already locked
+                return !automationPackageId.equals(byPassId);
+            } else {
+                throw new RuntimeException(BYPASS_AUTOMATION_PACKAGE_LOCK_FOR_ID + " is set in context but its value is not a string: " + byPassId);
+            }
+        }
+    }
+
     private static String getAutomationPackageId(ExecutionContext context) {
         Plan plan = context.getPlan();
         return (plan != null) ? (String) plan.getCustomField(AutomationPackageEntity.AUTOMATION_PACKAGE_ID):
                 null;
     }
 
     @Override
-    public void afterExecutionEnd(ExecutionContext context) {
-        super.afterExecutionEnd(context);
-        String automationPackageId = getAutomationPackageId(context);
-        if (automationPackageId != null) {
+    public void executionFinally(ExecutionContext context) {
+        super.executionFinally(context);
+        if (shouldLock(context)) {
```

Comment:
If for some reason the result of shouldLock would be different from the call in "executionStart" we would never release the lock. It might be safer to store the information that the lock has to be released within the context to be sure that we release it here

Reply:
I changed the implementation based on your comment